### ISSUE TITLE
A reentrant addCallbacks on a Deferred adds to mid-chain

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2849]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)| @Peaker reported|
|-|-|
|Trac ID|trac#2849|
|Type|defect|
|Created|2007-10-10 15:25:51Z|

Specifically, addCallbacks that executes from within the context of a callback, adds the callback after the current executing callback, rather than at the end of the chain.

<details><summary>Searchable metadata</summary>

```
trac-id__2849 2849
type__defect defect
reporter__Peaker Peaker
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1192029951000000 1192029951000000
changetime__1193953620000000 1193953620000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)|@Peaker removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_2"></a>|@Peaker commented|
|-|-|

Oops, the branch is: addcallbacks-2849

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_3"></a>|@jerub commented|
|-|-|

This code has tests, and the entire test suite runs without problems with the changes made. It worries me that this bug has slipped through.

I want to approve this commit, but I think that it's important to test some third party code just to make sure we're not going to break anyone else's codebase.

Perhaps run some of the Divmod test suite on this branch?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

I've run nevow test suite with success, so it looks good. My only complain would be to add documentation for ```_runningCallbacks```. But the final approval should go to our chief surgeon :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @Peaker|
|-|-|

Only looked at the code for a few seconds, but I noticed test_reentrantRunCallbacks doesn't return the Deferred it is playing with.  It's not really clear what the ZeroDivisionError there is helping to test (it certainly isn't documented).

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_7"></a>|@Peaker removed owner|
|-|-|

Thanks for the review. I documented the reason behind the ZeroDivisionError and returned the deferred.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)|@Peaker set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [21523]) better docstrings; reflow comments to proper width; other minor changes

refs [#2849](https://github.com/twisted/twisted/issues/2849)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [21524]) split the failure assertions into a separate test method

refs [#2849](https://github.com/twisted/twisted/issues/2849)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [21525]) update comment to talk about the new implementation

refs [#2849](https://github.com/twisted/twisted/issues/2849)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

I did the stuff mentioned above.  I think the branch looks fine now, although I noticed what might qualify as another bug while reviewing this.  I filed [#2861](https://github.com/twisted/twisted/issues/2861) for that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to @exarkun|
|-|-|

test_reentrantRunCallbacksWithFailure has an unused function callback2. In the same function, I think it would be more clear to have a separated function instead of a lambda followed by an assertEquals.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [21527]) remove unused nested function; define another one to assert about the resulting exception

refs [#2849](https://github.com/twisted/twisted/issues/2849)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve set owner to @exarkun|
|-|-|

Perfect, please merge!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set status to `closed`|
|-|-|

(In [21529]) Merge addcallbacks-2849

Author: peaker (mostly), exarkun
Reviewer: jerub, therve, exarkun
Fixes [#2849](https://github.com/twisted/twisted/issues/2849)

Fix Deferred.addCallback for the case where it is called on a Deferred from
a callback on that same Deferred.  The behavior is changed so that the new
callback is added to the end of the callback chain and only called after
the executing callback returns at the earliest.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
